### PR TITLE
Doc-service-accounts-How-to

### PR DESCRIPTION
**Pull request by bmckercher123**
New How to. Not tied to any release. Can merge at any time.

Reviewed by Alex.

### CODE REVIEW

**Comment by caseydavenport** [comment]
Deploy preview for *calico* ready!

Built with commit 6e5d3bbb881e648c39e4e01d2f9cfa8201972653

https://deploy-preview-2673--calico.netlify.com

File: master/security/service-accounts.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+### Concepts
+
+#### Use smallest set of permissions required
+
+Service accounts are controlled by RBAC so you can grant permissions to trusted entities (code and/or people). To perform any operation in a workload, clients are required to authenticate with the Kubernetes API server. If you do not explicitly assigned a service account to a pod, it uses the default ServiceAccount in the namespace. You should not grant broad permissions to the default service account for a namespace. If an application needs access to the Kubernetes API, create separate service accounts with the smallest set of permissions required. 
+
+### Before you begin...
+
+Configure Kubernetes service accounts with RBAC
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: Worth a link to the Kubernetes documentation for this?

> Configure Kubernetes service accounts with RBAC

Also, I don't think the RBAC part is relevant for this doc - really the prereq is just "Configure unique Kubernetes service accounts for your applications" or something like that, right?

File: master/security/service-accounts.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+### How to
+
+Like all other Kubernetes objects, service accounts have labels. So you can use labels to create ‘groups’ of service accounts. For Calico network policy, you can limit access to workloads using:
+
+- Matching endpoints in a service account; additionally, with a positive match on a namespace
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: "using matching endpoints in a service account" reads a bit odd. Maybe "_by_ matching endpoints in a service account"?

File: master/security/service-accounts.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+- Matching service account names and selectors
+
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: extra newline here

File: master/security/service-accounts.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+#### Limit ingress traffic for workloads by service account name
+
+In the following example, ingress traffic is limited to service accounts that match the names, **ingress-sa-in-calico-np** and **ingress-w-in-calico-np**.
+
+```
+apiVersion: projectcalico.org/v3
+kind: NetworkPolicy
+metadata: 
+  name: demo-calico-np
+  namespace: "prod-engineering"
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: no need for the quotes in the "namespace" field - it's automatically a string.

File: master/security/service-accounts.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        serviceAccounts: 
+          names: 
+            - ingress-sa-in-calico-np
+            - ingress-w-in-calico-np
+  selector: 'kind == "calico-np"'
+```
+
+#### Limit ingress traffic for workloads by service account label
+
+In the following example, ingress traffic is limited to services accounts that matches the label, **sa-in-calico-np**.
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: services -> service

"that matches" -> "that match"

Comment (caseydavenport) [reply master/security/service-accounts.md]: Also, a label is two parts - a key and a value. In this case, the key is "ingress" and the value is "sa-in-calico-np".

I'd reword this sentence to say: "In the following example, ingress traffic is allowed from any pods in the prod-engineering namespace whose service account has the label "ingress: sa-in-calico-np".

File: master/security/service-accounts.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+- Matching endpoints in a service account; additionally, with a positive match on a namespace
+- Matching service account names and selectors
+
+- [Limit ingress traffic for workloads by service account name](#limit-ingress-traffic-for-workloads-by-service-account-name)
+- [Limit ingress traffic for workloads by service account label](#limit-ingress-traffic-for-workloads-by-service-account-label)
+
+#### Limit ingress traffic for workloads by service account name
+
+In the following example, ingress traffic is limited to service accounts that match the names, **ingress-sa-in-calico-np** and **ingress-w-in-calico-np**.
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: I'd also reword this line - the key difference is that this policy doesn't limit traffic per-se, it allows a certain subset of traffic.

File: master/security/service-accounts.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  namespace: "prod-engineering"
+spec: 
+  ingress: 
+    - action: Allow
+      source: 
+        serviceAccounts: 
+          names: 
+          selector: 'ingress == "sa-in-calico-np"'
+  selector: 'kind == "calico-np"'
+
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: minor nit: extra newline here

Also, I'd consider changing the label names / values to be something a bit more realistic.

e.g. `selector: app == 'db'` and `selector: app == 'web-frontend'` or something like that, since I'd expect most deployments won't use a label like "kind: calico-np"

File: master/security/service-accounts.md (hunk 1 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 2 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 3 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: I see the two separate lists causes confusion so I moved this content up to Concepts.

File: master/security/service-accounts.md (hunk 4 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 5 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 7 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 6 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: _data/master/navbars/security.yml [hunk 8]
```diff
@@ -50,7 +50,11 @@ toc:
...
+- title: Use service accounts in policy rules
+  path: /security/service-accounts
 - title: Using HTTP methods and paths in policy rules
-  path: /security/http-methods 
+  path: /security/http-methods
 - title: Enable extreme high-connection workloads
-  path: /security/high-connection-workloads   
\ No newline at end of file
+  path: /security/high-connection-workloads   
+- title: Use service accounts in policy rules
```
Comment (caseydavenport) [inline _data/master/navbars/security.yml]: This shows up twice in the navbar (a few lines above as well)

File: master/security/service-accounts.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+Use Calico network policy to allow/deny traffic for Kubernetes services. 
+
+### Value
+
+Using Calico network policy, you can leverage Kubernetes service accounts with RBAC for flexible control over how policies are applied in a cluster. For example, the Network Security can have RBAC permissions to:
+
+- Control which service accounts the Developer team can use within a namespace
+- Write high-priority network policies for those service accounts (that the Developer team cannot override) 
+
+The Network Security team can maintain full control of security, while selectively allowing Developer operations where it makes sense.
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: I think "Network", "Security" and "Developer" should be lowercase in this sentence

Comment (caseydavenport) [reply master/security/service-accounts.md]: (and a couple of other places in the doc)

File: _data/master/navbars/security.yml (hunk 8 above)
Comment (bmckercher123) [reply _data/master/navbars/security.yml]: Fixed.

File: master/security/service-accounts.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+#### Use smallest set of permissions required
+
+Service accounts are controlled by RBAC so you can grant permissions to trusted entities (code and/or people). To perform any operation in a workload, clients are required to authenticate with the Kubernetes API server. If you do not explicitly assigned a service account to a pod, it uses the default ServiceAccount in the namespace. You should not grant broad permissions to the default service account for a namespace. If an application needs access to the Kubernetes API, create separate service accounts with the smallest set of permissions required. 
+
+#### Service account labels
+
+Like all other Kubernetes objects, service accounts have labels. So you can use labels to create ‘groups’ of service accounts. For Calico network policy, you can limit access to workloads by matching:
+
+- Endpoints in a service account (additionally, with a positive match on a namespace)
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: I think this list isn't quite right yet - "Endpoints in a service account" doesn't really make sense, since all of the options resolve down to that eventually.

I think the list should probably say this:

Calico network policy allows you to select workloads by their service account using either:
- an exact match on service account name
- a service account label selector expression

File: master/security/service-accounts.md [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+Configure unique Kubernetes service accounts for your applications
+
+### How to
+
+- [Limit ingress traffic for workloads by service account name](#limit-ingress-traffic-for-workloads-by-service-account-name)
+- [Limit ingress traffic for workloads by service account label](#limit-ingress-traffic-for-workloads-by-service-account-label)
+
+#### Limit ingress traffic for workloads by service account name
+
+In the following example, ingress traffic is allowed from any workload in the **prod-engineering** namespace, whose service account has the names **ingress-sa-in-calico** or **ingress-w-in-calico**.
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: A subtle point here on this phrase:

> ingress traffic is allowed from any workload in the **prod-engineering** namespace

It's actually not allowing from a specific namespace. Rather, it is allowing from any namespace with the label `ingress: prod-engineering`. 

The distinction is subtle, but it means this expression could select many namespaces, not just a single one, and those namespaces can be named whatever they like. The specific name of the namespace has no real relationship with the labels applied to that namespace.

As an example, I could have a namespace called "my-dev-namespace" and it could still have the label "ingress: prod-engineering" on it. In this case, it would be selected by this `namespaceSelector`.

File: master/security/service-accounts.md [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+kind: NetworkPolicy
+metadata: 
+  name: demo-calico
+  namespace: prod-engineering
+spec: 
+  ingress: 
+    - action: Allow
+      source: 
+        namespaceselector: 'ingress == "prod-engineering"'
+        selector: 'app == "db"'
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: I think let's drop the `selector` for this example - it's mean to be showcasing selection via service account, so the normal selector isn't required to show that.

File: master/security/service-accounts.md [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+kind: NetworkPolicy
+metadata: 
+  name: demo-calico
+  namespace: prod-engineering
+spec: 
+  ingress: 
+    - action: Allow
+      source: 
+        serviceAccounts: 
+          names:
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: We should remove `names:` from this example

File: master/security/service-accounts.md [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+  name: demo-calico
+  namespace: prod-engineering
+spec: 
+  ingress: 
+    - action: Allow
+      source: 
+        serviceAccounts: 
+          names: 
+          selector: 'app == "db"'
+  selector: 'app == "web-frontend"'
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: I'd swap the web-frontend label with the db one - it makes more sense that the web-frontend would be talking to the db than the other way around :)

File: master/security/service-accounts.md [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+spec: 
+  ingress: 
+    - action: Allow
+      source: 
+        namespaceselector: 'ingress == "prod-engineering"'
+        selector: 'app == "db"'
+        serviceAccounts: 
+          names: 
+            - ingress-sa-in-calico
+            - ingress-w-in-calico
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: What is `ingress-w-in-calico` mean to mean?

I think to make this example feel more realistic, these should be something like this:

```
names:
- api-service
- user-auth-service
```

The idea is that these service accounts aren't being created in order to be used in Calico policy - they will exist already for other purposes, but make great integration points with Calico policy (so, probably won't reference things "ingress" or "calico" in the name).

File: master/security/service-accounts.md [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+apiVersion: projectcalico.org/v3
+kind: NetworkPolicy
+metadata: 
+  name: demo-calico
+  namespace: prod-engineering
+spec: 
+  ingress: 
+    - action: Allow
+      source: 
+        namespaceselector: 'ingress == "prod-engineering"'
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: Similarly here, the namespace is unlikely to have a label with key `ingress` since that's a policy concept, and the namespace will exist outside of policy anyway.

It might have a label like this, though: `role: prod`

File: master/security/service-accounts.md (hunk 9 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: _data/master/navbars/security.yml (hunk 8 above)
Comment (bmckercher123) [reply _data/master/navbars/security.yml]: Fixed.

File: master/security/service-accounts.md (hunk 10 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 12 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 13 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 14 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 15 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Yeah, I just start with policy I can find in QA repo when I'm not sure how to create one. Good catch to make this more use case driven!

File: master/security/service-accounts.md (hunk 11 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Yikes, I see this was sloppy wording. Thanks for explaining this.

File: master/security/service-accounts.md (hunk 16 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed. Decided to keep it simple. Seems like it should be one example. Not sure what Alex had in mind. I'll go with whatever you think.

File: master/security/service-accounts.md [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+  name: demo-calico
+  namespace: prod-engineering
+spec: 
+  ingress: 
+    - action: Allow
+      source: 
+        serviceAccounts: 
+          names: 
+            - api-service
+            - use-auth-service
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: woops, I meant this to say `user-auth-service`

File: master/security/service-accounts.md [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+```
+
+#### Limit ingress traffic for workloads by service account label
+
+In the following example, ingress traffic is allowed from any workload whose service account matches the label, **app == web-frontend**.
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: I'd say "matches the label _selector_" here (subtle difference between a label and a selector - the thing written here is a selector rather than a label).

File: master/security/service-accounts.md [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+Configure unique Kubernetes service accounts for your applications
+
+### How to
+
+- [Limit ingress traffic for workloads by service account name](#limit-ingress-traffic-for-workloads-by-service-account-name)
+- [Limit ingress traffic for workloads by service account label](#limit-ingress-traffic-for-workloads-by-service-account-label)
+
+#### Limit ingress traffic for workloads by service account name
+
+In the following example, ingress traffic is allow from any workload whose service account matches the names **api-service** or **use-auth-service**.
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: "is allow" -> "is allowed"

File: master/security/service-accounts.md [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+title: Use service accounts in policy rules
+---
+
+### Big picture
+
+Use Calico network policy to allow/deny traffic for Kubernetes services. 
+
+### Value
+
+Using Calico network policy, you can leverage Kubernetes service accounts with RBAC for flexible control over how policies are applied in a cluster. For example, the network security can have RBAC permissions to:
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: > For example, the network security can have RBAC permissions to

"the network security" doesn't make sense here I don't think. What is this sentence meant to be saying?

File: master/security/service-accounts.md [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+
+- Control which service accounts the developer team can use within a namespace
+- Write high-priority network policies for those service accounts (that the developer team cannot override)
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: > - Control which service accounts the developer team can use within a namespace
> - Write high-priority network policies for those service accounts (that the developer team cannot override) 

Maybe impeded by the sentence above, but I don't know if this is right. Nothing about service account policies prevents developers from creating lower order (higher priority) network policies.

File: master/security/service-accounts.md [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+Use Calico network policy to allow/deny traffic for Kubernetes services. 
+
+### Value
+
+Using Calico network policy, you can leverage Kubernetes service accounts with RBAC for flexible control over how policies are applied in a cluster. For example, the network security can have RBAC permissions to:
+
+- Control which service accounts the developer team can use within a namespace
+- Write high-priority network policies for those service accounts (that the developer team cannot override) 
+
+The network security team can maintain full control of security, while selectively allowing developer operations where it makes sense.
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: Same here - I'm missing how this is connected to service accounts at all (or how this would be effectively done using Calico).

File: master/security/service-accounts.md [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+This how-to guide uses the following Calico features:
+
+**NetworkPolicy** or **GlobalNetworkPolicy** with a service account rule and match criteria.
+
+
+### Concepts
+
+#### Use smallest set of permissions required
+
+Service accounts are controlled by RBAC so you can grant permissions to trusted entities (code and/or people). To perform any operation in a workload, clients are required to authenticate with the Kubernetes API server. If you do not explicitly assign a service account to a pod, it uses the default ServiceAccount in the namespace. You should not grant broad permissions to the default service account for a namespace. If an application needs access to the Kubernetes API, create separate service accounts with the smallest set of permissions required.
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: Said this in a few PRs now - this line is too long. 

Notice how the GitHub UI treats this all as line 31. This should probably be no fewer than 4 distinct lines with explicit line returns in between. It makes it harder to use the GitHub UI to review when the lines are long, and also displays poorly in many common editors that won't add automatic breaks.

File: master/security/service-accounts.md (hunk 19 above)
Comment (caseydavenport) [reply master/security/service-accounts.md]: this still needs to be fixed.

File: master/security/service-accounts.md (hunk 17 above)
Comment (caseydavenport) [reply master/security/service-accounts.md]: Same here - this should be `user-auth-service`

File: master/security/service-accounts.md (hunk 18 above)
Comment (caseydavenport) [reply master/security/service-accounts.md]: ^ Comment still needs to be addressed

File: master/security/service-accounts.md [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+
+#### Limit ingress traffic for workloads by service account label
+
+In the following example, ingress traffic is allowed from any workload whose service account matches the label, **app == web-frontend**.
+
+```
+apiVersion: projectcalico.org/v3
+kind: NetworkPolicy
+metadata: 
+  name: demo-calico
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: To make this a more realistic policy, I think instead of `demo-calico` this should be called `allow-web-frontend`

File: master/security/service-accounts.md [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+#### Service account labels
+
+Like all other Kubernetes objects, service accounts have labels. You can use labels to create ‘groups’ of service accounts. Calico network policy lets you select workloads by their service account using:
+
+- An exact match on service account name
+- A service account label selector expression
+
+### Before you begin...
+
+Configure unique Kubernetes service accounts for your applications
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: Missing punctuation on this sentence.

File: master/security/service-accounts.md (hunk 9 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 17 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 18 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 19 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 21 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: This is Alex text.  It is just an example of how customers could use RBAC to control how policies are applied to a cluster. Not sure how to make this more clear...ideas?

File: master/security/service-accounts.md (hunk 23 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed. I have created separate lines, but this goes against all writing standards to break up sentences based on editor functions.

File: master/security/service-accounts.md (hunk 25 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 20 above)
Comment (caseydavenport) [reply master/security/service-accounts.md]: I think I understand the intention here now. How about something like this?

> Kubernetes RBAC allows you to control which users are allowed to create and use service accounts. Combined with Calico network policy, this allows for a security boundary between entities provisioning service accounts and those who are using them.

> For example, permission to provision new service accounts can be restricted using Kubernetes RBAC to only the network security team, who can then write high-priority network policies which reference those service accounts.

Comment (bmckercher123) [reply master/security/service-accounts.md]: I added this wording with some minor tweaks for clarity.

File: master/security/service-accounts.md [hunk 26]
```diff
@@ -0,0 +1,91 @@
+---
+title: Use service accounts in policy rules
+---
+
+### Big picture
+
+Use Calico network policy to allow/deny traffic for Kubernetes services.
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: I think this should say "using Kubernetes service accounts", right?

Otherwise this isn't a different "big picture" from our other policy doc?

File: master/security/service-accounts.md [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+In the following example, ingress traffic is allowed from any workload whose service account matches the names **api-service** or **user-auth-service**.
+
+```
+apiVersion: projectcalico.org/v3
+kind: NetworkPolicy
+metadata: 
+  name: demo-calico
+  namespace: prod-engineering
+spec: 
+  ingress:
```
Comment (caseydavenport) [inline master/security/service-accounts.md]: this guy is missing a selector like the example below has.

e.g., `selector: app == 'db'`

File: master/security/service-accounts.md (hunk 24 above)
Comment (caseydavenport) [reply master/security/service-accounts.md]: @bmckercher123 this comment is still valid ^

File: master/security/service-accounts.md (hunk 26 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Yes, I think that was there a long time without notice.

File: master/security/service-accounts.md (hunk 27 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.

File: master/security/service-accounts.md (hunk 20 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed

File: master/security/service-accounts.md (hunk 24 above)
Comment (bmckercher123) [reply master/security/service-accounts.md]: Fixed.
